docker/scripts/p4runtime_switch.py

```diff
@@ -0,0 +1,321 @@
+                 start_controller = True,
+                 program = None,
+                 verbose = False,
+                 device_id = None,
+                 enable_debugger = False,
+                 log_file = None,
+                 **kwargs):
+        Switch.__init__(self, name, **kwargs)
+        assert (sw_path)
+        self.sw_path = sw_path
+        # make sure that the provided sw_path is valid
+        pathCheck(sw_path)
+
+        self.program = program
+        json_path = None
+        p4info_path = None
+        if self.program:
+            json_path = self.program.json()
+            if self.program.p4info():
+                p4info_path = self.program.p4info()
+
+        if json_path is not None:
+            # make sure that the provided JSON file exists
+            if not os.path.isfile(json_path):
+                error("Invalid JSON file.\n")
+                exit(1)
+            self.json_path = json_path
+        else:
+            self.json_path = None
+
+        if p4info_path is not None:
+            if not os.path.isfile(p4info_path):
+                error("Invalid P4Info file.\n")
+                exit(1)
+            self.p4info_path = p4info_path
+        else:
+            self.p4info_path = None
+
+        if grpc_port is not None:
+            self.grpc_port = grpc_port
+        else:
+            self.grpc_port = P4RuntimeSwitch.next_grpc_port
+            P4RuntimeSwitch.next_grpc_port += 1
+
+        if thrift_port is not None:
+            self.thrift_port = thrift_port
+        else:
+            self.thrift_port = P4RuntimeSwitch.next_thrift_port
+            P4RuntimeSwitch.next_thrift_port += 1
+
+        if check_listening_on_port(self.grpc_port):
+            error('%s cannot bind port %d because it is bound by another process\n' % (self.name, self.grpc_port))
+            exit(1)
+
+        self.verbose = verbose
+        logfile = "/tmp/p4app-logs/p4s.{}.log".format(self.name)
+        self.output = open(logfile, 'w')
+        self.pcap_dump = pcap_dump
+        self.enable_debugger = enable_debugger
+        self.log_console = log_console
+        self.start_controller = start_controller
+        if not self.program.supportsP4Runtime():
+            self.start_controller = False
+        self.sw_conn = None
+        if log_file is not None:
+            self.log_file = log_file
+        else:
+            self.log_file = logfile
+        if device_id is not None:
+            self.device_id = device_id
+            P4Switch.device_id = max(P4Switch.device_id, device_id)
+        else:
+            self.device_id = P4Switch.device_id
+            P4Switch.device_id += 1
+        self.nanomsg = "ipc:///tmp/bm-{}-log.ipc".format(self.device_id)
+
+
+    def check_switch_started(self, pid):
+        for _ in range(SWITCH_START_TIMEOUT * 2):
+            if not os.path.exists(os.path.join("/proc", str(pid))):
+                return False
+            if check_listening_on_port(self.grpc_port):
```

Comment:
I kind of think that it's better to actually attempt the connection even if you have access to open ports

Reply:
The point of this code is to wait for the switch to start before attempting a connection. I think the original author of this code (@sethfowler ?) put it in for a reason.